### ISSUE TITLE
Enhancement: Avoid converting string literals to String

### ISSUE THREAD

**Issue by ryantrinkle** [ghcjs/ghcjs#599]
Even when using `JSString` via `OverloadedStrings`, it appears that string literals get converted to `String` and then back.  This seems to cause a fairly substantial amount of work on application startup; it would be great to avoid this, if possible.

**Comment by daig** [comment]
Is the `JSSTRING pack -> fused` rule firing correctly?

You may be able to coerce a `"primitiveString"# :: Addr#` via `String.fromCharCode.apply` as a workaround.

Another (simple but extremely ugly) alternative is to use template haskell for the performance critical strings like:
```
js_literal s =
return [ForeignD (ImportF JavaScript Unsafe ("$r = '" ++ s ++ "'") (mkName ("literal_" ++ s)) (ConT ''JSString))]
```
Which should theoretically have the lowest overhead but I haven't tested either solution extensively.

Personally I think `OverloadedStrings` should always be handled at compile-time as an upstream change to GHC, so it may be worth pursuing the elegant solution there rather than hacking a special case into GHCJS.

**Comment by luite** [comment]
I once started working on this with the intention to make this work, using RULES, but I don't quite remember where exactly I stopped or if there were major blockers.

I agree that most of this should be handled by GHC.

`OverloadedStrings` is currently implemented by inserting a `fromString` call, and then hoping that the optimiser can do the right thing with that. Perhaps that's enough (but in that case it'd be nice to have some kind of warnings for leftover `fromString` applications)

Still, the GHC implementation of strings is rather C/native code specific: literals get translated to `decodeCString` on some packed version of the data. That's a bit awkward to work with. Maybe a bit of customisation of the generated STG would make things easier.

**Comment by dmjio** [comment]
@luite in the meantime, would it be better to not use `OverloadedStrings` and just manually `pack` things? So as to get around converting to/from `String` twice.

**Comment by ryantrinkle** [comment]
This seems to work:

```haskell
foreign import javascript unsafe "$r = $1;" addrToByteArray :: Addr# -> ByteArray#

{-# RULES
"inlineJSString2" forall s. T.pack (GHC.Base.build (unpackFoldrCString# s)) = Text (JSString (JSVal (addrToByteArray s)))
  #-}
```

**Comment by luite** [comment]
That looks more specific to the `Text` wrapper than I'd like, but something along those lines would be nice to have indeed.

But does this really completely do away with all conversions? STG doesn't have a way to represent the string literal as a JS String, and while the code generator has some hacks to get rid of unnecessary conversions, I'd still be a little surprised if there's no `h$str` or similar call somewhere.

**Comment by ryantrinkle** [comment]
@luite I'm not sure if it gets rid of them completely in all cases, but it did seem to in the cases I looked at.  I've added this hack to the `Text` wrapper library: https://github.com/obsidiansystems/text/pull/8

**Comment by luite** [comment]
Ok, I'll review this.

I'm glad that it works, but I'm a bit worried that this is somewhat unintentional and possibly fragile.

**Comment by ryantrinkle** [comment]
Yeah, it would definitely be nice to have something more solid.  Any idea what that might look like?  This definitely impacts startup time, which is pretty important.
